configure

```diff
@@ -0,0 +1,265 @@
+# something non-standard.
+#
+# Important environment variables:
+#
+#     BUILD_PATH - path to build directory (default: "build")
+#
+# ./configure expects to be run from the source root with build directory
+# configured by BUILD_PATH if necessary.
+#
+# Run
+#
+#     ./configure --help
+#
+# to see available options.
+
+set -eu
+
+BUILD_PATH=${BUILD_PATH:-build}
+
+CC=${CC:-cc}
+CMAKE=${CMAKE:-cmake}
+
+usage() {
+    cat <<EOF
+Usage:
+    $0 [options...] [<var>=<value>...]
+
+Options:
+    --prefix=path           installation prefix
+    --bindir=path           installation path for binaries
+    --includedir=path       installation path for headers
+    --libdir=path           installation path for shared libraries
+    --jnidir=path           installation path for JNI libraries
+    --pkgconfigdir=path     installation path for pkg-config files
+
+    --help                  read this help
+EOF
+}
+
+die() {
+    exec >&2
+    echo $@
+    echo
+    usage
+    exit 1
+}
+
+configure_mk="$BUILD_PATH/configure.mk"
+
+write_boilerplate() {
+    mkdir -p "$(dirname "$configure_mk")"
+    cat > "$configure_mk"  <<EOF
+# DO NOT EDIT
+#
+# Generated automatically by $0
+
+EOF
+}
+
+set_variable() {
+    eval "$1=\"$2\""
+    echo "$1 = $2" >> "$configure_mk"
+}
+
+append_variable() {
+    eval "$1=\"$1 $2\""
+    echo "$1 += $2" >> "$configure_mk"
+}
+
+standard_dirs="prefix bindir includedir libdir jnidir pkgconfigdir"
+
+is_standard_dir() {
+    for dir in $standard_dirs
+    do
+        if [ "$1" = "$dir" ]
+        then
+            return 0
+        fi
+    done
+    return 1
+}
+
+parse_variable_override() {
+    case "$1" in
+      --*=*)
+        var=${1#--}
+        value=${var#*=}
+        var=${var%%=*}
+        if is_standard_dir "$var"
+        then
+            set_variable "$var" "$value"
+            nshift=1
+            return 0
+        fi
+        die "unrecognized option: $1"
+        ;;
+
+      --*)
+        var=${1#--}
+        if is_standard_dir "$var"
+        then
+            if [ $# = 1 ]
+            then
+                die "option $1 expects an argument"
+            fi
+            value=$2
+            set_variable "$var" "$value"
+            nshift=2
+            return 0
+        fi
+        die "unrecognized option: $1"
+        ;;
+
+      *=*)
+        var=${1%%=*}
+        value=${1#*=}
+        set_variable "$var" "$value"
+        nshift=1
+        return 0
+        ;;
+    esac
+    return 1
+}
+
+parse_commandline() {
+    while [ $# -gt 0 ]
+    do
+        if [ "$1" = "--help" ]
+        then
+            usage
+            exit
+        fi
+
+        if parse_variable_override "$@"
+        then
+            shift $nshift
+            continue
+        fi
+
+        die "unrecognized argument: $1"
+    done
+}
+
+detect_platform() {
+    case "$(uname)" in
+      Darwin)
+        set_variable IS_MACOS "true"
+        ;;
+      Linux)
+        set_variable IS_LINUX "true"
+        ;;
+      *)
```

Comment:
I'd suggest:
* to specify explicitly pattern for MS: `MINGW*|CYGWIN*|MSYS*)`
* to throw an exception on unknown platforms

Reply:
> specify explicitly pattern for MS

IIRC, this one is specifically for MSYS2, not a generic Windows platform. Windows support is in kinda "purgatory" state, since it's not tested automatically, and not regularly tested manually either. So I'd be wary to add new behavior without proper testing.

I'd suggest to keep the current behavior (detect only MSYS2) and add other platforms later, in a more systematic way.

> throw an exception on unknown platforms

Hm... Seems like a good idea, *but* it may break some builds. Currently, if we cannot detect the platform we assume it to be generic UNIX system. This way `make` will probably work on various BSDs, etc. If we disallow unknown systems then we might break builds for some users that build Themis for platforms that we don't explicitly support yet they still work.

Are we okay with that?

Reply:
Sounds reasonable. Thanks.